pkg/didcomm/common/service/service.go

```diff
@@ -35,17 +35,44 @@ type Header struct {
 	Type   string           `json:"@type"`
 }
 
+func (h *Header) clone() *Header {
+	if h == nil {
+		return nil
+	}
+
+	return &Header{
+		ID: h.ID,
+		Thread: decorator.Thread{
+			ID: h.Thread.ID,
+		},
+		Type: h.Type,
+	}
+}
+
 // DIDCommMsg did comm msg
 type DIDCommMsg struct {
 	Header  *Header
 	Payload []byte
 }
 
+// Clone creates new DIDCommMsg with the same data
+// the cloned message is safe for delivering to the client
+// it prevents modifying by the client
+func (m *DIDCommMsg) Clone() DIDCommMsg {
+	if m == nil {
+		return DIDCommMsg{}
+	}
+	return DIDCommMsg{
+		Header:  m.Header.clone(),
+		Payload: append(m.Payload[:0:0], m.Payload...),
+	}
+}
+
 // NewDIDCommMsg returns DIDCommMsg with Header
-func NewDIDCommMsg(payload []byte) (*DIDCommMsg, error) {
-	msg := &DIDCommMsg{Payload: payload}
+func NewDIDCommMsg(payload []byte) (DIDCommMsg, error) {
+	msg := DIDCommMsg{Payload: payload}
 	if err := json.Unmarshal(msg.Payload, &msg.Header); err != nil {
-		return nil, fmt.Errorf("invalid payload data format: %w", err)
+		return msg, fmt.Errorf("invalid payload data format: %w", err)
```

Comment:
This shouldn't return non-nil err value with a non-nil val ie `return val, err`.

Reply:
yeah, but it is impossible, we can return nil only if `DIDCommMsg` is a pointer

Reply:
ok. I will revert it to pointer

Reply:
Done!